### PR TITLE
Abstracted robot collision checking into class for snap and vectorfield

### PR DESCRIPTION
**Pull request by cdellin**
In the push for baked checking in PrPy, this is a first pass which creates a two robot collision checkers:
- `SimpleRobotCollisionChecker`
- `BakedRobotCollisionChecker`
Each of these can be constructed for a robot, and expose a simple `CheckCollision` method, which raises a PrPy exception upon collision.

This PR also refactors the `SnapPlanner` and `VectorFieldPlanner` planners to make use of this abstraction.  The planners default to `SimpleRobotCollisionChecker`, but a different type can be passed in when they are instantiated, and this type will be used for its collision checking.

Note that the `BakedRobotCollisionChecker` is experimental, and only works with the `baked_checkers` branch of `or_fcl` at the moment.  Currently, the underlying checker does not expose which links provoked the collision, so the collision report is not helpful, and a vanilla `CollisionPlanningError` is currently thrown.

I'll wait to implement this for `TSRPlanner` until its PR (#322) is merged.

### CODE REVIEW

File: src/prpy/collision.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import openravepy
+from prpy.planning.exceptions import CollisionPlanningError
+from prpy.planning.exceptions import SelfCollisionPlanningError
+
+
+class SimpleRobotCollisionChecker:
```
Comment (mkoval) [inline src/prpy/collision.py]: Missing docstring.

File: src/prpy/collision.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        elif self.robot.CheckSelfCollision(report=report):
+            raise SelfCollisionPlanningError.FromReport(report)
+
+
+class BakedRobotCollisionChecker:
+
+    def __init__(self, robot):
+        self.robot = robot
+        self.env = robot.GetEnv()
+        self.checker = self.env.GetCollisionChecker()
```
Comment (mkoval) [inline src/prpy/collision.py]: Can you check the type of `GetCollisionChecker` before trying to use it?

File: src/prpy/collision.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        kb_type = self.checker.SendCommand('BakeGetType')
+        self.checker.SendCommand('BakeBegin')
+        self.env.CheckCollision(self.robot)
+        self.robot.CheckSelfCollision()
+        self.baked = openravepy.RaveCreateKinBody(self.env, kb_type)
```
Comment (mkoval) [inline src/prpy/collision.py]: Check for an error here.

File: src/prpy/collision.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        self.checker.SendCommand('BakeEnd')
+
+    def CheckCollision(self):
+        report = openravepy.CollisionReport()
+        if self.checker.CheckSelfCollision(self.baked, report):
```
Comment (mkoval) [inline src/prpy/collision.py]: Please add a comment explaining that this performs the baked check.

File: src/prpy/collision.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+class BakedRobotCollisionChecker:
+
+    def __init__(self, robot):
+        self.robot = robot
+        self.env = robot.GetEnv()
+        self.checker = self.env.GetCollisionChecker()
+        kb_type = self.checker.SendCommand('BakeGetType')
+        self.checker.SendCommand('BakeBegin')
+        self.env.CheckCollision(self.robot)
+        self.robot.CheckSelfCollision()
```
Comment (mkoval) [inline src/prpy/collision.py]: Please add a comment explaining that this bakes the two collision checks.

File: src/prpy/planning/snap.py [hunk 6]
```diff
@@ -36,6 +36,7 @@
     ClonedPlanningMethod,
     Tags
 )
+from prpy.collision import SimpleRobotCollisionChecker
```
Comment (mkoval) [inline src/prpy/planning/snap.py]: Prefer: `from ..collision import SimpleRobotCollisionChecker`

File: src/prpy/planning/vectorfield.py [hunk 7]
```diff
@@ -35,6 +35,7 @@
 import openravepy
 from .. import util
 from base import BasePlanner, PlanningError, ClonedPlanningMethod, Tags
+from prpy.collision import SimpleRobotCollisionChecker
```
Comment (mkoval) [inline src/prpy/planning/vectorfield.py]: Prefer: `from ..collision import SimpleRobotCollisionChecker`

**Comment by mkoval** [comment]
Looks good, other than the minor comments I left above. Can you also add this to CHOMP, GreedyIk, and TSRPlanner?

File: src/prpy/collision.py (hunk 1 above)
Comment (cdellin) [inline src/prpy/collision.py]: Added docstrings.

File: src/prpy/collision.py (hunk 2 above)
Comment (cdellin) [inline src/prpy/collision.py]: I don't check the type, but I do now check for the case where no checker exists, and I raise a more helpful exception when the checker doesn't support baked checking.

File: src/prpy/collision.py (hunk 3 above)
Comment (cdellin) [inline src/prpy/collision.py]: Once we've confirmed that the checker supports baking, this is assumed to work.  If not, it already throws an exception (kinbody type not known!) which is correct.

File: src/prpy/collision.py (hunk 4 above)
Comment (cdellin) [inline src/prpy/collision.py]: Comment added!

File: src/prpy/collision.py (hunk 5 above)
Comment (cdellin) [inline src/prpy/collision.py]: Comment added.

File: src/prpy/planning/snap.py (hunk 6 above)
Comment (cdellin) [inline src/prpy/planning/snap.py]: Fixed (against my better judgment `(-:`).

**Comment by cdellin** [comment]
I addressed most of the comments, and also added this for `ChompPlanner`.

I'd prefer to wait to add this to `TSRPlanner` until #322 is merged.  Also, the current implementation of `GreedyIKPlanner` does its collision checking in `FindIKSolutions`, and unfortunately there's no way to add baked support to that planner.

**Comment by cdellin** [comment]
I like how immediately after my last comment, you went ahead and merged #322.  Fine!  `TSRPlanner` support is now in.  (I refactored the robot collision interface to make it a bit nicer and more consistent with the native OpenRAVE interface.)

File: src/prpy/collision.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        self.baked = openravepy.RaveCreateKinBody(self.env, kb_type)
+        self.checker.SendCommand('BakeEnd')
+
+    def CheckCollision(self, report=None):
+        # The baked check is performed by checking self collision on baked
+        if self.checker.CheckSelfCollision(self.baked, report):
+            return True
+        return False
```
Comment (mkoval) [inline src/prpy/collision.py]: Replace this with:
```python
return self.checker.CheckSelfCollision(self.baked, report)
```
:grin:

File: src/prpy/planning/tsr.py [hunk 9]
```diff
@@ -136,12 +138,16 @@ def compute_ik_solutions(tsrchain):
 
             return ik_solutions
 
+        # Instantiate a robot checker
+        with CollisionOptionsStateSaver(
+                self.env.GetCollisionChecker(), CollisionOptions.ActiveDOFs):
+            robot_checker = self.robot_collision_checker(robot)
+
```
Comment (mkoval) [inline src/prpy/planning/tsr.py]: You should move the `SetActiveDOFs` call above the baking.

File: src/prpy/collision.py (hunk 8 above)
Comment (cdellin) [inline src/prpy/collision.py]: Haha done!

File: src/prpy/planning/tsr.py (hunk 9 above)
Comment (cdellin) [inline src/prpy/planning/tsr.py]: Good catch! Fixed.

**Comment by cdellin** [comment]
Comments addressed!

**Pull request merged**